### ISSUE TITLE
`apm dedupe` shouldn't require Visual Studio

### ISSUE THREAD

**Issue by rivy** [atom/apm#347]
After installing Atom and the [`atom-beautify`](https://github.com/Glavin001/atom-beautify) package into an otherwise bare machine, I'm left with some exceedingly long file paths in the Atom package folder. These long paths are causing some of my other software (notably OneDrive) some difficulties.

So, to repair the issue, I attempted to use `apm dedupe` to condense the path, but I've run into this error...

```
C:\Users\Roy\AppData\Local\atom\app-0.194.0\resources\app\apm\lib\dedupe.js:94
        throw new Error('You must have Visual Studio 2010, 2012, or 2013 insta
              ^
Error: You must have Visual Studio 2010, 2012, or 2013 installed
    at Dedupe.module.exports.Dedupe.getVisualStudioFlags (C:\Users\Roy\AppData\Local\atom\app-0.194.0\resources\app\apm\lib\dedupe.js:94:15)
    at Dedupe.module.exports.Dedupe.forkDedupeCommand (C:\Users\Roy\AppData\Local\atom\app-0.194.0\resources\app\apm\lib\dedupe.js:120:25)
    at Dedupe.module.exports.Dedupe.dedupeModules (C:\Users\Roy\AppData\Local\atom\app-0.194.0\resources\app\apm\lib\dedupe.js:100:19)
    at C:\Users\Roy\AppData\Local\atom\app-0.194.0\resources\app\apm\lib\dedupe.js:162:24
    at fn (C:\Users\Roy\AppData\Local\atom\app-0.194.0\resources\app\apm\node_modules\async\lib\async.js:582:34)
    at Object._onImmediate (C:\Users\Roy\AppData\Local\atom\app-0.194.0\resources\app\apm\node_modules\async\lib\async.js:498:34)
    at processImmediate [as _immediateCallback] (timers.js:354:15)
```

It seems incorrect and more than a bit off-the-wall that Visual Studio would be a requirement to perform the deduplication of the apm/npm modules.

Note that while I'll open another issue against [`atom-beautify`](https://github.com/Glavin001/atom-beautify) requesting simplification of their module path structure, this is an `apm` error. I haven't had the time to look deeply into the code to understand the stated requirement, but it should be removed.

Thanks for your attention.

**Comment by lee-dohm** [comment]
I suspect if you have native-compiled modules as part of the set of dependencies that Visual Studio may well be a requirement for installation of those packages, let alone deduping of the dependencies. You could ask that a package not depend on something that requires native-compiled code ... but that may mean that performance of that package would suffer.

So if a compiler is a requirement for installation of some packages, is requiring a complier for deduping still an issue?

**Comment by rivy** [comment]
I appreciate your thoughts on the matter. And, I agree, if a compiler is needed to install a package, it seems reasonable to require it for a dedupe operation (although I can see ways around that).

In this case, however, no compiler was used for installation of any of the atom packages. All of the packages were installed using the usual Atom install via the Settings page. No compiler was ever on this PC. So, I don't see a reason for a Visual Studio requirement error. It's definitely a defect. I'm just not sure where to lay the blame yet.

**Comment by lee-dohm** [comment]
If all installed packages didn't require a compiler to install, then yes, it shouldn't be required for deduping them. I'm just making certain that you weren't objecting to the concept of a compiler being required *at all*.

I'm also concerned that we'll keep running into this problem with npm modules and Windows "long path" problem. Because from what I can tell npm installs dependencies of dependencies of dependencies as `[installation path]\node_modules\foo\node_modules\baz\node_modules\...` *ad infinitum*.

**Comment by rivy** [comment]
Yep, windows definately has an ugly problem (arguably, a full bug) working with long pathnames.

UNC pathnames can be used to work around the woefully short (~250 character) path limitation, but many current tools (including windows` CMD) are not, or are only partially compatible, with longer pathnames.

And the `npm` module installation / dependency strategy rams into that issue squarely. There's been some talk about the npm module issue (http://stackoverflow.com/questions/26155135/node-npm-windows-file-paths-are-too-long-to-install-packages, https://github.com/npm/npm/issues/5641, https://github.com/joyent/node/issues/6960#issuecomment-46704998), and I know lots of folks are thinking about it. But it's not clear from the discussions that a  resolution is coming anytime soon.

In the meanwhile, `apm dedupe` is a workaround which might work, except I'm getting this error on multiple Windows machines. I can find a copy of Visual Studio to install, but that seems excessive for what seems to be a spurious error.

What's the best strategy to get this issue fixed?

**Comment by kevinsawicki** [comment]
Sorry about this, this was meant to be removed in https://github.com/atom/apm/pull/141

Will push up a fix shortly.

**Comment by kevinsawicki** [comment]
This will be fixed in the next Atom release 0.195, thanks for reporting it.

**Comment by lee-dohm** [comment]
:+1:
